client/login.go

```diff
@@ -0,0 +1,186 @@
+	CodeVerifier = cv
+	CodeVerifierMux.Unlock()
+	fmt.Printf("Code verifier set: %v\n", CodeVerifier)
+	w.WriteHeader(http.StatusOK)
+}
+
+// Handles the exchange of the authorization code for tokens. This is the URL that the server redirects the browser to after the user logs in.
+func consumerExchangeHandler(w http.ResponseWriter, r *http.Request) {
+	queryParams := r.URL.Query()
+	authCode := queryParams.Get("code")
+	redirectURL := *Server + "/oauth-landing/"
+	if authCode == "" {
+		http.Error(w, "Authorization Failed. Authorization code was not provided.", http.StatusBadRequest)
+		return
+	}
+
+	responseJSON, status, error := GetTokens(authCode, "")
+	if status == -1 {
+		text := "Authorization Failed. Server is not reachable. Response: " + error
+		http.Error(w, text, http.StatusBadRequest)
+		return
+	}
+
+	if status != 200 {
+		text := fmt.Sprintf("Authorization Failed. Retrieval of tokens failed (status code: %d) Response: %v", status, error)
+		http.Error(w, text, status)
+		return
+	}
+
+	TasksMux.Lock()
+	for appID := range Tasks {
+		taskID := uuid.New().String()
+		task := NewTask(make(map[string]interface{}), appID, taskID, "login")
+		task.Result = responseJSON
+		task.Finish("Tokens obtained")
+		Tasks[appID][task.TaskID] = task
+	}
+	TasksMux.Unlock()
+
+	http.Redirect(w, r, redirectURL, http.StatusPermanentRedirect)
+}
+
+// GetTokens sends a request to the server to get tokens. It returns the response JSON, status code and error message as string.
+// Parameter authCode is the authorization code - if it's not empty, it's used to get the tokens in grant_type "authorization_code".
+// Parameter refreshToken is the refresh token - if it's not empty, it's used to get the tokens in grant_type "refresh_token".
+// Must be called with either authCode or refreshToken, not both.
+func GetTokens(authCode string, refreshToken string) (map[string]interface{}, int, string) {
+	data := url.Values{}
+	var grantType string
+	if authCode != "" {
+		grantType = "authorization_code"
+		data.Set("code", authCode)
+		CodeVerifierMux.Lock()
+		defer CodeVerifierMux.Unlock()
+		if CodeVerifier != "" {
+			data.Set("code_verifier", CodeVerifier)
+		} else {
+			return nil, -1, "Could not find code_verifier."
+		}
+	} else if refreshToken != "" {
+		grantType = "refresh_token"
+		data.Set("refresh_token", refreshToken)
+	} else {
+		return nil, -1, "No authCode or refreshToken provided"
+	}
+
+	data.Set("grant_type", grantType)
+	data.Set("client_id", OAUTH_CLIENT_ID)
+	data.Set("scopes", "read write")
+	data.Set("redirect_uri", fmt.Sprintf("http://localhost:%s/consumer/exchange/", *Port))
+
+	url := fmt.Sprintf("%s/o/token/", *Server)
+	req, err := http.NewRequest("POST", url, strings.NewReader(data.Encode()))
+	if err != nil {
+		log.Fatalf("Error creating request: %v", err)
+	}
+
+	req.Header = getHeaders("", *SystemID, "", "")                      // Request comes from the browser, so we cannot set addonVersion and platformVersion
```

Comment:
Well, in the case of refresh_token, you can ;)
In the case of authorization_code, you could use the `state` parameter to encode the versions in the [authorization request](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1). You would get it in the `state` parameter of the [authorization response](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2) then and use it here if desired :)